### PR TITLE
Add a doc and script for grpc compilation

### PR DESCRIPTION
**Pull request by yashsingh74**
This PR has the document about how to build the gRPC stubs and the script to generate them.

Fixes for #282

### CODE REVIEW

**Comment by mcluseau** [comment]
hi, I think the protobuf subdir will break the package we want. To me the command should be something like

    protoc -I ./ --go_out=paths=source_relative:. --go-grpc_out=paths=source_relative:. $(find -name '*.proto')

**Comment by yashsingh74** [comment]
> hi, I think the protobuf subdir will break the package we want. To me the command should be something like
> 
> ```
> protoc -I ./ --go_out=paths=source_relative:. --go-grpc_out=paths=source_relative:. $(find -name '*.proto')
> ```

@micahhausler 
I ran that command from the home directory of this project and gives me the following result-
```
protoc -I ./ --go_out=paths=source_relative:. --go-grpc_out=paths=source_relative:. $(find -name '*.proto')
find: illegal option -- n
usage: find [-H | -L | -P] [-EXdsx] [-f path] path ... [expression]
       find [-H | -L | -P] [-EXdsx] -f path [path ...] [expression]
Missing input file.
```

**Comment by rajaskakodkar** [comment]
```
protoc -I ./ --go_out=paths=source_relative:. --go-grpc_out=paths=source_relative:. $(find . -name '*.proto')
```
cc: @yashsingh74 this worked for me. Also, try to fix the make target for modd here https://github.com/kubernetes-sigs/kpng/blob/master/modd.conf.template#L5

@jayunit100  @knabben  do we need scripts under hack/ or can we make the make target with modd work?

**Comment by anusha94** [comment]
> @jayunit100 @knabben do we need scripts under hack/ or can we make the make target with modd work?

And maybe need another Good First Issue to have dependent targets that will install `protoc`?
Until then, having the scripts makes sense. But maybe not once we get `make modd` working

**Comment by aahemm** [comment]
Hi @yashsingh74,
I am kind of a beginner in this topic. I have gone through 
your commits and have a couple of questions: a) Shouldn't the verify script exit with a non zero exit code when the pb files differ? It will make the pipeline fail and I think this is the desired behavior. b) Where do we generate the files in the `examples` directory? I mean the files that are `diff`ed against actual pb files.

File: examples/protobuf/services.pb.go [hunk 1]
```diff
@@ -0,0 +1,1992 @@
+// Copyright 2021 The Kubernetes Authors.
```
Comment (anusha94) [inline examples/protobuf/services.pb.go:1]: we should not check-in `examples` dir. This is only to verify the diff with the generated one vs the checked-in one.
Also, I think a better name to this dir would be something like `verify` or `validate`

Comment (yashsingh74) [reply examples/protobuf/services.pb.go:1]: Where should I keep the verify dir ?

Comment (jayunit100) [reply examples/protobuf/services.pb.go]: hack/verify... is fine @yashsingh74

**Comment by jayunit100** [comment]
thank you for this yash ! lets try to merge it this wk

**Comment by jayunit100** [comment]
i think im ok w/ this ? it doesnt break anything and its a step in the right direction.  that said i think probably we should work together on this so that its clear what the trajectory is.

if nobody objects ill merge it tomorrow.

**Comment by jayunit100** [comment]
@yashsingh74 can you rebase to one commit please?

**Comment by mcluseau** [comment]
I'm approving the tests to run, but am I the only one who sees the github button to do that? @jayunit100

**Comment by mcluseau** [comment]
@yashsingh74 thanks for this PR, it even has doc so it's better than what I do :p a little scripts needs its `chmod +x` though: test fails with `./hack/verify/generate-grpc.sh: Permission denied`

**Comment by yashsingh74** [comment]
@mcluseau I have make the script executable. Please try around once more.

**Comment by mcluseau** [comment]
@yashsingh74 ok now it runs, but still fails (protoc dependency missing) https://github.com/kubernetes-sigs/kpng/runs/7427413370?check_suite_focus=true

**Comment by jayunit100** [comment]
unit tests + grpc yay ! ok , lets get them passing now @yashsingh74

File: hack/verify/verify-grpc.sh [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Verifying the protobuf
+
+function verify {
+    diff -q api/localnetv1/verify-protobuf/services.pb.go api/localnetv1/services.pb.go && diff -q api/localnetv1/verify-protobuf/services_grpc.pb.go api/localnetv1/services_grpc.pb.go || { echo "File doesn't match"; exit 1; }
```
Comment (jayunit100) [inline hack/verify/verify-grpc.sh:20]: can we print the entire diff?

File: .github/workflows/grpc_tests.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        go version
+        go install golang.org/x/lint/golint@latest
+        sudo apt-get install -y libnl-3-dev libnl-genl-3-dev
+
+    - name: checkout
+      uses: actions/checkout@v2
+     
+    - name: run grpc verify tests
+      run: |
+        ./hack/verify/generate-grpc.sh
```
Comment (jayunit100) [inline .github/workflows/grpc_tests.yml:33]: `./hack/verify/generate-grpc.sh && ./hack/verify/verify-grpc.sh` ?

**Comment by jayunit100** [comment]
talked w/ @mcluseau @yashsingh74 

Lets 
- add Dockerfile into hack/ 
- use *that* to generate the protobuf
- use `ubuntu:latest` probably as the Dockerfile? 
- file a follow on issue to make as many jobs as we can, 
```
jobs:
  test:
    name: run grpc verify tests
```
run in a container (the same one ideally?) so that theres 100% parity between the ci jobs build env and developer env.

**Comment by jayunit100** [comment]
@anusha94 do you want to help yash finish this one ?

**Pull request closed without merging**